## Assignment #1

---

* Release date: 2020/04/13
* Due date: **2020/04/19 23:59** (will not accept late submission)
* Submittion format: notebook file which can be executed in Colab environment
* Weighting: 5% (total 50 pts)

---

1. **(10pts)** Calculate `np.dot(x, rotation)` and explain how this calculation is performed.

```python
  import numpy as np

  x = np.array([2, 0])
  rotation = np.array([ [0, -1],
                        [1,  0] ])
```

In [0]:
import numpy as np

x=np.array([2,0])
rotation=np.array([[0, -1],
                   [1, 0]])

np.dot(x,rotation)

# array([ 0 -2 ])
# the value of theta is 90. So it rotates points in the xy plane anticlockwise through an angle 90 about the origin.

array([ 0, -2])

출력 되는 값: array([0, -2])

이유: The value of theta is 90(sin90 = 1, cos90 = 0). So it rotates points in the xy plane anticlockwise through an angle 90 about the origin.



2. **(10pts)** Suppose we have the following 2D tensor (i.e., a matrix). How to rearrange its values into 1D tensor (i.e., a vector) in a column major order?
```python
x = np.array([[1,  2,  3,  4],
                 [5,  6,  7,  8],
                 [9, 10, 11, 12]])
```

In [0]:
x=np.array([[1,2,3,4],
            [5,6,7,8],
            [9,10,11,12]])

x=x.reshape((12,))
print(x)

[ 1  2  3  4  5  6  7  8  9 10 11 12]


3. **(10pts)** Compute a transpose of the matrix `x` in Problem 2 by using only `np.reshape` function.

In [0]:
x=np.array([[1,2,3,4],
            [5,6,7,8],
            [9,10,11,12]])

x=x.reshape(-1,order='F')
x=x.reshape(4,3)
print(x)

[[ 1  5  9]
 [ 2  6 10]
 [ 3  7 11]
 [ 4  8 12]]


열(column)에 대해 순서가 적용되게 reshape 해주고, 이를 다시 4 by 3 행렬로 reshape 해준다.





4. **(10pts)** Given any random matrices, compute the element-wise multiplication using a naive Python implementation and Numpy built-in function respectively. Compare the wall-clock times of these implementations as the size of matrices increases.

In [0]:
#multiplication using naive Python implementation

import numpy as np
import time

start = time.time()

def naive_vector_dot(x,y):
  assert len(x.shape)==1
  assert len(y.shape)==1
  assert x.shape[0]==y.shape[0]

  z=0
  for i in range(x.shape[0]):
    z += x[i] * y[i]
  return z

def naive_matrix_vector_dot(m,n,r):
  x = np.random.randint(10,size=(m,n))
  y = np.random.randint(10,size=(n,r))
  assert len(x.shape) == 2
  assert len(y.shape) == 2
  assert x.shape[1] == y.shape[0]

  z=np.zeros((x.shape[0], y.shape[1]))
  for i in range(x.shape[0]):
    for j in range(y.shape[1]):
      row_x=x[i, :]
      column_y=y[:,j]
      z[i,j]=naive_vector_dot(row_x, column_y)
  return z


A=naive_matrix_vector_dot(10,50,15)
print(A)
print('the time of multiply (10 by 50) and (50 by 15) matrices : %.4f' % (time.time()-start))

B=naive_matrix_vector_dot(100,500,150)
print(B)
print('the time of multiply (100 by 500) and (500 by 150) matrices : %.4f' % (time.time()-start))

[[ 892.  975. 1267. 1178. 1013.  960. 1065.  972. 1140. 1182. 1120.  824.
  1102.  958. 1000.]
 [ 941. 1002. 1090. 1077.  983.  864. 1154.  914. 1098. 1000.  866.  725.
  1056.  909.  811.]
 [ 857.  966. 1154.  987.  907.  905. 1083. 1106. 1091. 1000. 1015.  737.
  1010. 1007.  831.]
 [ 894.  931. 1081. 1044.  895.  924. 1034. 1062. 1198.  986. 1104.  730.
  1104.  962.  932.]
 [ 820.  992. 1142. 1196.  963. 1022. 1049.  998. 1128.  964. 1005.  728.
  1117. 1074.  894.]
 [ 897.  895. 1018. 1027. 1046.  960. 1030. 1058. 1169. 1215.  996.  816.
  1102. 1036.  956.]
 [ 764. 1151. 1137. 1209. 1051.  862. 1247.  997. 1079.  974. 1062.  752.
   986.  989.  765.]
 [ 834. 1042. 1023. 1106. 1075.  834. 1151. 1073. 1208. 1028. 1034.  822.
  1199.  989.  835.]
 [ 888. 1074. 1111. 1104. 1071.  971. 1189. 1127. 1235. 1103. 1065.  816.
  1243. 1088. 1018.]
 [ 888.  974. 1170. 1129. 1041. 1037. 1036. 1119. 1246. 1065. 1169.  819.
  1171. 1111.  963.]]
the time of multiply (10 by 50) and (50 by 15) ma

naive Python 으로 했을 때

(10 by 50), (50 by 15) 매트릭스를 연산하면 0.0058초가 걸리고,

(100 by 500), (500 by 150) 매트릭스를 연산하면 3.5267초가 걸린다는 것을 알 수 있다.

즉, 행렬의 크기가 커질수록 결과를 내는 데에 시간이 더 많이 소요된다.




In [0]:
#multiplication using Numpy built-in function

import numpy as np
import time

start=time.time()

def naive_matrix_vector_dot(m,n,r):
  x = np.random.randint(10,size=(m,n))
  y = np.random.randint(10,size=(n,r))
  z=np.dot(x,y) # 행렬의 곱셈 연산을 하려면 x.shape[1] == y.shape[0] 이어야 한다.
  return z

A=naive_matrix_vector_dot(10,50,15)
print(A)
print('the time of multiply (10 by 50) and (50 by 15) matrices : %.4f' % (time.time()-start))

B=naive_matrix_vector_dot(100,500,150)
print(B)
print('the time of multiply (100 by 500) and (500 by 150) matrices : %.4f' % (time.time()-start))

[[ 999  968 1097 1028 1096 1057 1094  948  920 1191 1072 1107  821 1228
  1276]
 [1085 1142 1007 1053 1163 1123 1016 1022  989 1280 1046 1113 1196 1373
  1261]
 [1144 1181 1345 1203 1238 1233 1147 1169 1095 1484 1187 1366 1098 1510
  1488]
 [ 939  845 1075 1056 1047 1078  949  890  926 1099  943 1031  945 1223
  1147]
 [ 842  895  974  926 1046 1020  946  864  797  972  806  980  892 1100
  1070]
 [ 927  944 1004  906 1034  961  912  796  845 1100  935  970  941 1154
  1067]
 [ 939  993 1075 1048 1032 1000  992  830  888 1132  970 1044  921 1173
  1083]
 [ 868  911 1118  983  934  988  950  827  964 1095  913  972  988 1128
  1128]
 [1035 1105 1047 1134 1026 1148 1069 1011  998 1293  998 1296 1042 1251
  1350]
 [ 977 1030 1108  989 1050 1117 1009  993  955 1197 1020 1110  940 1264
  1183]]
the time of multiply (10 by 50) and (50 by 15) matrices : 0.0023
[[ 9814 10158  9421 ... 10180 10041 10404]
 [10370 10568  9862 ... 10582 10166 10590]
 [10305 10289  9557 ... 10373  9836 10179]
 ...


Numpy built-in function 으로 했을 때

(10 by 50), (50 by 15) 매트릭스를 연산하면 0.0023초가 걸리고,

(100 by 500), (500 by 150) 매트릭스를 연산하면 0.0187초가 걸린다는 것을 알 수 있다.

결국 크기가 큰 매트릭스의 연산에 소요되는 시간을 비교 해보면, Numpy built-in function이 naive Python에 비해 훨씬 더 빠르게 처리된다는 것을 알 수 있다.



5. **(10pts)** Consider MNIST classification problem covered during the class. For the details, please refer to [this material](https://drive.google.com/a/ds.seoultech.ac.kr/file/d/1lWLwZQGXgZJq4PFHvgra6D6U420ra-O3/view?usp=sharing). In this example, we used the multilayer perceptron composed of an input layer with 512 hidden nodes and an output layer that produces predicted probabilities over 10 classes. In the class, we used GPU as a hardware accelerator to train our model.

  Here, let's verify the actual benefit of using GPU for training. For this, compare the wall-clock times in the case of 1) using CPU and 2) using GPU to train MNIST classifier.

In [9]:
import tensorflow as tf
tf.__version__

import time

from tensorflow.keras.datasets import mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

from tensorflow.keras import models
from tensorflow.keras import layers

network = models.Sequential()
network.add(layers.Dense(512, activation='relu', input_shape=(28*28,)))
network.add(layers.Dense(10, activation='softmax'))

network.compile(optimizer='rmsprop', loss='categorical_crossentropy', metrics=['accuracy'])

train_images=train_images.reshape((60000, 28*28))
train_images=train_images.astype('float32') / 255

test_images=test_images.reshape((10000, 28*28))
test_images=test_images.astype('float32') / 255

from tensorflow.keras.utils import to_categorical

train_labels=to_categorical(train_labels)
test_labels=to_categorical(test_labels)

start=time.time()

with tf.device('/CPU:0'):
  nf=network.fit(train_images, train_labels, epochs=5, batch_size=128)

run_time_cpu = time.time()-start
test_loss, test_acc = network.evaluate(test_images, test_labels)
print('test_acc: ', test_acc)
print('Elapsed time of using CPU: %.4f' % (run_time_cpu))

start1=time.time()

with tf.device('/GPU:0'):
  nf=network.fit(train_images, train_labels, epochs=5, batch_size=128)

run_time_gpu = time.time()-start1
test_loss, test_acc = network.evaluate(test_images, test_labels)
print('test_acc: ', test_acc)
print('Elapsed time of using GPU: %.4f' % (run_time_gpu))

print('GPU speedup over CPU: %.4f'%(run_time_cpu/run_time_gpu))


Epoch 1/5
469/469 [==============================] - 3s 7ms/step - loss: 0.2519 - accuracy: 0.9275
Epoch 2/5
469/469 [==============================] - 3s 7ms/step - loss: 0.1030 - accuracy: 0.9698
Epoch 3/5
469/469 [==============================] - 4s 8ms/step - loss: 0.0673 - accuracy: 0.9799
Epoch 4/5
469/469 [==============================] - 3s 7ms/step - loss: 0.0483 - accuracy: 0.9855
Epoch 5/5
313/313 [==============================] - 1s 3ms/step - loss: 0.0731 - accuracy: 0.9777
test_acc:  0.9776999950408936
Elapsed time of using CPU: 17.4259
Epoch 1/5
469/469 [==============================] - 2s 4ms/step - loss: 0.0285 - accuracy: 0.9914
Epoch 2/5
469/469 [==============================] - 2s 4ms/step - loss: 0.0222 - accuracy: 0.9934
Epoch 3/5
469/469 [==============================] - 2s 4ms/step - loss: 0.0169 - accuracy: 0.9950
Epoch 4/5
469/469 [==============================] - 2s 4ms/step - loss: 0.0130 - accuracy: 0.9963
Epoch 5/5
313/313 [=========================

with tf.device 로 연산을 수행할 장치(CPU/GPU)를 직접 선택한다.

'Elapsed time of using CPU: 17.4259초

'Elapsed time of using GPU: 8.8967초

위의 코드 실행결과로, GPU로 실행했을 때가 CPU로 실행했을 때보다 1.9587배 더 빠르게 처리된다는 것을 알 수 있다.


